File: Go/Linked List/61_Rotate List.go
import "math"

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil || k == 0 {
		return head
	}

	temp := head
	var count int = 1
	for temp.Next != nil {
		temp = temp.Next
		count += 1
	}
	if k > count {
		k = k % count
	}
	reqCount := int(math.Abs(float64(count - k)))
	if reqCount == 0 || k == 0 {
		return head
	}
	temp = head
	count = 1
	for temp.Next != nil {
		if count == reqCount {
			break
		}
		temp = temp.Next
		count += 1
	}
	headOrg := head
	head = temp.Next
	temp.Next = nil
	temp = head
	for temp.Next != nil {
		temp = temp.Next
	}
	temp.Next = headOrg
	return head
}

File: C++/Linked List/61_Rotate_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *rotateRight(ListNode *head, int k)
    {
        ListNode *dummy = new ListNode(0, head);
        ListNode *first = dummy, *second = dummy;
        ListNode *tempHead = head;
        int size = 0;
        while (tempHead != NULL)
        {
            tempHead = tempHead->next;
            size++;
        }
        if (size == 0 || size == 1 || size == k)
            return head;
        if (k > size)
        {
            k = k % (size);
        }
        for (int i = 1; i <= k + 1 && first != NULL; i++)
        {
            first = first->next;
        }
        while (first != NULL)
        {
            first = first->next;
            second = second->next;
        }
        ListNode *temp = second;
        while (second->next != NULL)
        {
            second = second->next;
        }
        second->next = head;
        head = temp->next;
        temp->next = NULL;
        return head;
    }
};
